packages/pubsub/src/publisher.js

```diff
@@ -0,0 +1,179 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*!
+ * @module pubsub/publisher
+ */
+
+'use strict';
+
+var arrify = require('arrify');
+var common = require('@google-cloud/common');
+var extend = require('extend');
+var is = require('is');
+
+/**
+ *
+ */
+function Publisher(topic, options) {
+  this.topic = topic;
+  this.api = topic.api;
+
+  this.inventory = {
+    queued: [],
+    inFlight: []
+  };
+
+  Object.defineProperty(this.inventory, 'queueSize', {
+    get: function() {
+      return this.queued.reduce(function(size, message) {
+        return size + message.data.length;
+      }, 0);
+    }
+  });
+
+  this.settings = extend(true, {
+    batching: {
+      maxBytes: Infinity,
+      maxMessages: Infinity,
+      maxMilliseconds: 1000
+    }
+  }, options);
+
+  this.publishHandle_ = null;
+}
+
+/**
+ *
+ */
+Publisher.prototype.cancelPublish_ = function() {
+  clearTimeout(this.publishHandle_);
+  this.publishHandle_ = null;
+};
+
+/**
+ *
+ */
+Publisher.prototype.prepareMessages_ = function() {
+  var size = 0;
+  var messages = [];
+
+  var message;
+
+  while (message = this.inventory.queued.shift()) {
+    var isUnderSizeLimit = (size + message.data.length) <=
+      this.settings.batching.maxBytes;
+
+    var isUnderMessageLimit = (messages.length + 1) <=
+      this.settings.batching.maxMessages;
+
+    if (!isUnderSizeLimit || !isUnderMessageLimit) {
+      this.inventory.queued.unshift(message);
+      break;
+    }
+
+    messages.push(message);
+  }
+
+  return messages;
+};
+
+/**
+ *
+ */
+Publisher.prototype.publish = function(data, attrs, callback) {
+  if (!(data instanceof Buffer)) {
+    throw new Error('Data must be in the form of a Buffer.');
+  }
+
+  if (is.fn(attrs)) {
+    callback = attrs;
+    attrs = {};
+  }
+
+  this.inventory.queued.push({
+    data: data,
+    attrs: attrs,
+    callback: callback
+  });
+
+  var reachedMaxMessages = this.inventory.queued.length >=
+    this.settings.batching.maxMessages;
+
+  var reachedMaxBytes = this.inventory.queueSize >=
+    this.settings.batching.maxBytes;
+
+  // publish immediately if we hit maxMessage or maxBytes
+  if (reachedMaxMessages || reachedMaxBytes) {
+    this.publish_();
+  } else if (!this.timeoutHandle_) {
+    this.schedulePublish_();
+  }
+};
+
+/**
+ *
+ */
+Publisher.prototype.publish_ = function() {
+  var self = this;
+
+  this.cancelPublish_();
+
+  var messages = this.prepareMessages_();
+
+  var reqOpts = {
+    topic: this.topic.name,
+    messages: messages.map(function(message) {
+      return {
+        data: message.data,
+        attributes: message.attrs
+      };
+    })
+  };
+
+  this.inventory.inFlight = messages;
```

Comment:
I'm a PubSub person. Please don't cap the number of outgoing requests. There's lots of problems with this:

1) It breaks the other limits in the batching settings. If a user says "a batch must have foo extra latency, contain at most bar messages, be at most baz bytes", you won't be able to honor those limits at the same time as capping outgoing requests
2) It puts a ceiling on our publish throughput equal to max concurrent requests * max batch size. We don't want to put a ceiling on publish throughput. We originally proposed including publisher flow control in the library, and after much discussion, agreed **not** to have it. But this is essentially publisher flow control.
3) Waiting for concurrent requests to complete will increase rather than decrease system stability. Imagine there's a blip in Pub/Sub's availability such that the in-flight publish calls are lost and will timeout. Do you want to smear that blip across a longer period of time, or do you want to absorb it and move on? The ability to publish should be a function of present batching state and present Pub/Sub availability... it should not depend on past Pub/Sub availability.